lib/signs/bus.ex

```diff
@@ -0,0 +1,111 @@
+defmodule Signs.Bus do
+  use GenServer
+  require Logger
+
+  @enforce_keys [
+    :id,
+    :pa_ess_loc,
+    :text_zone,
+    :audio_zones,
+    :sources,
+    :config_engine,
+    :prediction_engine
+  ]
+  defstruct @enforce_keys
+
+  def start_link(sign, opts \\ []) do
+    state = %__MODULE__{
+      id: Map.fetch!(sign, "id"),
+      pa_ess_loc: Map.fetch!(sign, "pa_ess_loc"),
+      text_zone: Map.fetch!(sign, "text_zone"),
+      audio_zones: Map.fetch!(sign, "audio_zones"),
+      sources:
+        for source <- Map.fetch!(sign, "sources") do
+          %{
+            stop_id: Map.fetch!(source, "stop_id"),
+            routes:
+              for route <- Map.fetch!(source, "routes") do
+                %{
+                  route_id: Map.fetch!(route, "route_id"),
+                  direction_id: Map.fetch!(route, "direction_id")
+                }
+              end
+          }
+        end,
+      config_engine: opts[:config_engine] || Engine.Config,
+      prediction_engine: opts[:prediction_engine] || Engine.BusPredictions
+    }
+
+    GenServer.start_link(__MODULE__, state)
+  end
+
+  def init(state) do
+    schedule_run_loop(self())
+    {:ok, state}
+  end
+
+  def handle_info(:run_loop, state) do
+    schedule_run_loop(self())
+
+    %__MODULE__{
+      id: id,
+      sources: sources,
+      config_engine: config_engine,
+      prediction_engine: predictions_engine
+    } = state
+
+    _config = config_engine.sign_config(id)
+    current_time = Timex.now()
+
+    _predictions =
+      for %{stop_id: stop_id, routes: routes} <- sources,
+          prediction <- predictions_engine.predictions_for_stop(stop_id),
+          Enum.any?(
+            routes,
+            &(&1.route_id == prediction.route_id && &1.direction_id == prediction.direction_id)
+          ) do
+        prediction
+      end
+      # Exclude predictions that are too old
+      |> Enum.reject(&Timex.before?(&1.departure_time, Timex.shift(current_time, seconds: -5)))
+      # Special case: exclude 89.2 OB to Davis
+      |> Enum.reject(&(&1.stop_id == "5104" && String.starts_with?(&1.headsign, "Davis")))
+      # Special case: exclude routes terminating at Braintree (230.4 IB, 236.3 OB)
+      |> Enum.reject(&(&1.stop_id == "38671" && String.starts_with?(&1.headsign, "Braintree")))
+      # Special case: exclude routes terminating at Mattapan, in case those variants of route 24 come back.
+      |> Enum.reject(
+        &(&1.stop_id in ["185", "18511"] && String.starts_with?(&1.headsign, "Mattapan"))
+      )
```

Comment:
We could make this a bit more efficient with the use of [Streams](https://hexdocs.pm/elixir/1.14.3/Stream.html). Each function in the `Enum` module more or less iterates over the entire collectable and creates intermediate lists between each call. Streams are lazy and composable so would save us 4 iterations over the list of predictions in this case. Luckily, their APIs are very similar, if not identical. Check out https://hexdocs.pm/elixir/1.14.3/Stream.html#reject/2!

Reply:
Done, thanks for the pointer.